### PR TITLE
Task/gps and location sharing/received message handler

### PR DESCRIPTION
**Pull request by teamagam1**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/11" target="_blank">ReviewNinja

File: app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessagePollingService.java [hunk 1]
```diff
@@ -123,9 +126,14 @@ private void handleActionMessagePolling() {
...
         Log.v(LOG_TAG, "MessagePolling service processing " + messages.size() + " new messages");
 
-        for (Message m :
+        for (Message msg :
                 messages) {
-            MessagePubSub.getInstance().publish(m);
+            if (msg.getSenderId().equals(NetworkUtil.getMac())) {
+                continue;
+            }
+            Intent intent = new Intent(msg.getClass().getName());
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessagePollingService.java]: these 3 lines can be a static method in MessageBroadcastReceiver:
static void broadcastMessage(Message msg)

File: app/src/main/java/com/teamagam/gimelgimel/app/utils/GsonUtil.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import com.teamagam.gimelgimel.app.model.ViewsModels.Message;
+import com.teamagam.gimelgimel.app.model.ViewsModels.MessageJsonAdapter;
+
+/**
+ * Created on 5/2/2016.
+ * mainly used for {@link MessageJsonAdapter}
+ */
+public class GsonUtil {
+
+    private static final Gson gson = new GsonBuilder()
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/utils/GsonUtil.java]: rename to sGosn to meet conventions !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java [hunk 3]
```diff
@@ -210,31 +227,22 @@ protected void onPostCreate(Bundle savedInstanceState) {
...
-                    public void run() {
-                        ShowMessageDialogFragment.showNewMessages(getFragmentManager(), msg);
-                    }
-                });
-            }
-        };
-        MessagePubSub.getInstance().subscribe(subscriber);
+        // Register to receive messages.
+        // We are registering an observer
+        LocalBroadcastManager.getInstance(this).registerReceiver(
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java]: can also be a static method at MessageBroadcastReceiver:
register(MessageBroadcastReceiver mbr)

File: app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java [hunk 4]
```diff
@@ -210,31 +227,22 @@ protected void onPostCreate(Bundle savedInstanceState) {
...
+                mLatLongMessageReceiver, mLatLongMessageReceiver.getIntentFilter());
 
         mLocationFetcher.registerForUpdates(5000, 0);
     }
 
     @Override
     protected void onPause() {
         super.onPause();
+        LocalBroadcastManager.getInstance(this).unregisterReceiver(mTextMessageReceiver);
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java]: static unregistration method.

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+@Config(constants = BuildConfig.class, sdk = 21)
+public class GsonUtilTest {
+
+    @Before
+    public void setUp() throws Exception {
+
+    }
+
+    @Test
+    public void testGsonUtil_fromJson_shouldBeEqual() throws Exception {
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: should be seperated into different tests. one for text-messages and the other for LatLongMessages

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        //Act
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java:49]: This is actualy the //Arrange part

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        String senderId = "sender1";
+        String text = "text123";
+
+        Message msgT = new MessageText(senderId, text);
+        Message msgL = new MessageLatLong(senderId, new PointGeometry(23, 32));
+        Message[] messages = new Message[]{msgT, msgL};
+
+
+        for (Message msg : messages) {
+            String msgJson = GsonUtil.toJson(msg);
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: //Act

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            Message msgObj = GsonUtil.fromJson(msgJson, Message.class);
+            assertEquals(msgObj.getClass(), msg.getClass());
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: //Assert

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 5 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: This method checks both toJson and fromJson.
There should be one for each method

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Created on 5/2/2016.
+ */
+
+@RunWith(RobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, sdk = 21)
+public class GsonUtilTest {
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: can be removed if not used

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 8 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fix

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 7 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fix

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 6 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java:49]: !fix

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 5 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fix

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 9 above)
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fix

File: app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+
+    @Before
+    public void setUp() throws Exception {
+        mShadowContext = spy(ShadowApplication.getInstance().getApplicationContext());
+        mMessageHandler = mock(MessageBroadcastReceiver.NewMessageHandler.class);
+    }
+
+    @Test()
+    public void testConstructor_AllMessages_shouldCreateAll() throws Exception {
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java:43]: a test per message type

File: app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, sdk = 21)
+public class MessageBroadcastReceiverTest {
```
Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java:29]: Seperation into arrange, act and assert should be used

Comment (BarAgam) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java:29]: !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessagePollingService.java (hunk 1 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessagePollingService.java]: !resolve

Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessagePollingService.java]: !resolved

File: app/src/main/java/com/teamagam/gimelgimel/app/utils/GsonUtil.java (hunk 2 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/utils/GsonUtil.java]: !fixed

File: app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java (hunk 3 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java]: !resolved

File: app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java (hunk 4 above)
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/view/MainActivity.java]: !resolved

File: app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java (hunk 11 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/model/ViewsModels/MessageBroadcastReceiverTest.java:29]: !fixed

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 9 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fixed

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 5 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fixed

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 6 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java:49]: !fixed

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 7 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fixed

File: app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java (hunk 8 above)
Comment (teamagam1) [inline app/src/test/java/com/teamagam/gimelgimel/app/utils/GsonUtilTest.java]: !fixed

**Pull request merged**